syntax = "proto3";

//api.[cqrs].[tag].[version]
package api.query.webclient.v1;

import "google/protobuf/empty.proto";
//import "google/protobuf/timestamp.proto";
import "google/protobuf/field_mask.proto";

service Profile {

  rpc Get (ProfilesRequest) returns (ProfileDataList);
  rpc Identification (google.protobuf.Empty) returns (ProfileData);

}

message ProfileData {
  string createdAt = 1;
  string nickname = 2;
  string email = 3;
  string description = 4;
}

message ProfileDataList {
  repeated ProfileData profiles = 1;
}

//--- Request

message ProfilesRequest {
  ProfileDataList profilesData = 1;
  google.protobuf.FieldMask fields = 2;
}



service Record {
  
  rpc Get(RecordsPaginationRequest) returns (RecordsCursorResponse);

}

message RecordData {
  string id = 1;
  string createdAt = 2;
  string lastUpdate = 3;
  string nickname = 4;
  string record = 5;
}

message RecordDataList {
  repeated RecordData records = 1;
}

//---

message RecordsCursorResponse {
  RecordDataList recordList = 1;
  string cursor = 2;
}

//---

message EditRecordRequest {
  RecordData record = 1;
  google.protobuf.FieldMask fields = 2;
}

enum RecordKey {
  UNKNOWN = 0;
  ID = 1;
  CREATEDAT = 2;
  LASTUPDATE = 3;
  NICKNAME = 4;
  RECORD = 5;
}

enum Operand {
  NONE = 0;
  IN = 1; // in or ==
  OUT = 2; // out or !=
  LESS = 3; // less or <
  GREATER = 4; // greater or >
  EQLESS= 5; // not greater or <=
  EQGREATER = 6; // not less or >=
}

enum Direction {
  NOT = 0;
  MIN = 1;
  MAX = 2;
}

message Filter {
  RecordKey key = 1;
  Operand operand = 2;
  repeated string values = 3;
}

message Ordering {
  RecordKey key = 1;
  Direction direction = 2; 
}

message RecordsPaginationRequest {
  string cursor = 1;

  repeated Filter filters = 2;

  repeated Ordering orders = 3;

  int32 limit = 4;
}